lib/commander/runner.rb

```diff
@@ -341,8 +341,12 @@ def remove_global_options(options, args)
         past_switch, arg_removed = false, false
         args.delete_if do |arg|
           break if arg == '--'
+
+          arg_equals_index = arg.index('=') if switch_has_arg
+          arg = arg[0, arg_equals_index] if arg_equals_index
+
           if switches.any? { |s| s[0, arg.length] == arg }
-            arg_removed = !switch_has_arg
+            arg_removed = !switch_has_arg || arg_equals_index
```

Comment:
We deem the argument removed if there was an `=` in the argument. ie the switch name and switch value is contained in the one argument. No need to remove the subsequent argument.

Reply:
The wording/naming here reads in a pretty confusing way to me (it already did, but let's see if we can improve it). Maybe it's the name `switch_has_arg` — should that really be `any_switch_has_arg` ? And then `!switch_has_arg || arg_equals_index` really means something like "no switches have args, or, this arg has an argument and matches the given switch" which could be reflected in a comment for clarity. Probably part of the confusion here is that we're referring to two different things as `arg`. May be out of scope for this PR but it would be great to make that less confusing at some point.

Reply:
I agree, it's pretty complex.

Regarding `switch_has_arg`. I think changing this to `option_has_argument` would help: if any of the option's declared switches has an argument defined, then all the switches for that option require an argument.

I've had a go at renaming and refactoring. How does this look? Should I add it to the pull request or propose it in a subsequent one?

```ruby
    def remove_global_options(options, args)
      options.each do |option|
        switches = option[:switches]
        next if switches.empty?
        option_has_argument = switches.any? { |s| s =~ /[ =]/ }
        switches = expand_optionally_negative_switches(switches)

        option_argument_needs_removal = false
        args.delete_if do |token|
          break if token == '--'

          # Use just the portion of the argument before the = when
          # comparing switches.
          index_of_equals = token.index('=') if option_has_argument
          token = token[0, index_of_equals] if index_of_equals
          token_contains_option_argument = !index_of_equals.nil?

          if switches.any? { |s| s[0, token.length] == token }
            option_argument_needs_removal = option_has_argument && !token_contains_option_argument
            true
          elsif option_argument_needs_removal && token !~ /^-/
            option_argument_needs_removal = false
            true
          else
            false
          end
        end
      end
    end
```

Renaming as follows:

- `switch_has_arg` -> `option_has_argument`
- `arg` -> `token`
- `arg_equals_index` -> `index_of_equals`

Introduced:

- `token_contains_option_argument`

Also `past_switch` and `arg_removed` have been combined into one flag: `option_argument_needs_removal`. This makes the line in question more clear IMO.

```ruby
option_argument_needs_removal = option_has_argument && !token_contains_option_argument
```

Reply:
I like the renaming, go ahead and include it in this PR. Thanks!

Reply:
Thanks. Added in 008f0ad.